File: 백준/Gold/10026. 적록색약/적록색약.py
import sys
input = sys.stdin.readline
from collections import deque
N = int(input())
arr = [list(input().strip()) for _ in range(N)]


directions  = [
    [0, -1],[0,1],
    [1,0],[-1,0]
]
answer = [0, 0] ## 정상인, 적록 색맹

## 그냥 두번 돌려도 됨
def bfs(r,c, isSick): ## r, c, 적록색맹 여부 
    que = deque()
    color = arr[r][c]
    que.append([r,c]) ## r, c, type
    visited[r][c] = True
    while que:
        r,c = que.popleft()
        for [dr, dc] in directions:
            nr, nc = r+dr, c+dc
            if(not(0<=nr<N) or not(0<=nc<N) ) or visited[nr][nc]: ## 범위 나간 경우
                continue
             ## 적록 색맹인 경우
            if isSick == True and ((color=="B" and arr[nr][nc]=="B") or (color!="B" and arr[nr][nc]!="B")):
                    visited[nr][nc]=True
                    que.append([nr, nc])
            elif isSick==False and color == arr[nr][nc]:
                    visited[nr][nc]= True
                    que.append([nr, nc])

for idx in range(len(answer)):
    isSick = False if idx==0 else True
    visited = [[False for _ in range(N)] for _ in range(N)]
    for i in range(N):
        for j in range(N):
            if not(visited[i][j]):
                answer[idx]+=1
                bfs(i, j , isSick)

print(*answer)

File: 백준/Gold/10026. 적록색약/적록색약.js
const readline = require("readline")
const rl = readline.createInterface({
    input: process.stdin,
    output: process.stdout
})

const lines = []
const directions = [
    [0,-1],[0,1],
    [1,0],[-1,0]
]

function bfs(r,c, isSick, visited, graph,N){
    const que = [[r,c]]
    const color = graph[r][c]
    visited[r][c]=true
    while (que.length>0){
        const [r,c] =  que.shift()
        for (let [dr, dc] of directions){
            const [nr, nc] = [r+dr, c+dc]
            // 범위 나감, 방문함
            if (!(0<=nr && nr<N) || !(0<=nc && nc<N) || (visited[nr][nc])){
                continue
            }
            if (isSick && ((color =="B" && graph[nr][nc]=="B") || (color!="B" && graph[nr][nc]!="B"))){
                visited[nr][nc] = true
                que.push([nr, nc])                
            }
            else if (!isSick && (graph[nr][nc] ===color)){
                visited[nr][nc] = true
                que.push([nr, nc])
            }
        }
    }
}
    

rl.on("line", (line) => {

    lines.push(line.trim())

}).on("close", () => {

    const N = Number(lines[0])
    const graph = lines.slice(1).map(e => e.split(""))
    // console.log(N, graph)
    const answer = [0,0]


    for (let idx = 0; idx<answer.length;idx++){
        const isSick = idx==0?false: true
        const visited = Array.from(Array(N), ()=> Array(N).fill(false))
        for(let i=0;i<N;i++){
            for(let j=0; j<N;j++){
                if(!visited[i][j]){
                    answer[idx]+=1
                    bfs(i, j, isSick, visited, graph,N)
                }
                
            }
        }
    }
    console.log(answer.join(" "))
})